### ISSUE TITLE
Recommended way to get Packer built VM ID into Terraform control

### ISSUE THREAD

**Issue by Blefish** [bpg/terraform-provider-proxmox#1412]
**Is your feature request related to a problem? Please describe.**
I am using Packer to automate VM templates. My current issue is how to reference the newly built VM ID in Terraform to create new virtual machines based off the template. Ideally I'd build multiple VM templates (perhaps once a week) and allow it to iterate without pain when creating new VMs. Previous VM templates may live a while before they get cleaned up and all environments have switched over.

**Describe the solution you'd like**
I was wondering if there would be a way to request file/snippets data source. This way, it would be possible to store the VM ID in Proxmox files and reference it later from Terraform, always getting up-to-date VM template.

Or perhaps filter VM templates by age, similar to what AWS `data.aws_ami` offers via `latest = true` . I'd imagine this may not be so straight forward though, as VM template is very similar to a regular VM. This could be used together with data source `data.proxmox_virtual_environment_vms`

**Describe alternatives you've considered**
My current alternative is to use manifest JSON created by Packer after build and parse that into Terraform. While it works for local development, for integration with CI, it would make more sense to store the file outside repo. GitLab has snippets & CI/CD variables for this via API, which could be used as well.

**Additional context**
Just wondering if maybe there is already a better way to solve this problem via Proxmox :)

Possibly related: https://github.com/bpg/terraform-provider-proxmox/issues/1396

**Comment by bpg** [comment]
Hi @Blefish 👋🏼

I'm wondering if the [`virtual_environment_vms`](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/data-sources/virtual_environment_vms) data source would be any help in your use case? You can assign a unique tag to the Packer-generated template VM, then query the VM by tag using this data source and get the VM ID from the result.

**Comment by Blefish** [comment]
My main issue is how to figure out which one is the latest. I somehow need to publish this information and read it back.

I'm not sure if Proxmox exposes a timestamp when the VM was first created, but if that would be available, it would also solve the problem. Or I could read and write it to the VM template description instead, which could also work.

I think that other methods would all require storing the information outside the VM resource in that sense, which would be also fine and I could write a wrapper around it to receive it. For example, during packer build, it would store the VM ID in proxmox files storage/gitlab/github variables and then in terraform, I could read it out and always have the latest at hand. However perhaps others have also stumbled upon this and found a better workaround.

Basically I am looking for a similar functionality that is offered by AWS `most_recent` which automatically sorts it. I'm fine with manually finding out the latest one but currently lacking the information to do so.

```terraform
data "aws_ami" "amazon_linux" {
  most_recent = true
  owners      = ["amazon"]

  filter {
    name   = "name"
    values = ["al2023-ami-minimal-*"]
  }
}
```

**Comment by vanillaSprinkles** [comment]
yep tags likely the best method for the time being; then pull the tags list and sort with some locals - example of using tags for this towards the bottom


i dont think the time-created for vms, or time-converted-to-template is tracked in each vm-id's payload (at least not via pvesh) - we may want to request this on proxmox itself (if not already on their task list, i have not checked that far - came up simply with https://forum.proxmox.com/threads/api-to-read-qemu-vm-creation-time-or-uptime.45092/)

there is a roundabout method via pvesh: i was able to query the cluster tasks for type `qmtemplate` and that gives fields `starttime` and `endtime` and the vm id; might be the path needed (but id hope proxmox would include timestamps directly in the vm# payload vs logs). i am hoping a task-list per vm exists, but currently i think it requires pulling all tasks then filtering


example commands from my test system:


`pvesh get /nodes/anode/qemu/101/config --output-format json-pretty`

``` json
{
   "boot" : "order=scsi0;ide2;net0",
   "cores" : 1,
   "cpu" : "x86-64-v2",
   "digest" : "f818ee9fac93fba534893d266580465d17190370",
   "ide2" : "local-btrfs:iso/archlinux-2024.03.29-x86_64.iso,media=cdrom,size=978252K",
   "memory" : "2048",
   "meta" : "creation-qemu=8.1.5,ctime=1716443232",
   "name" : "arch01-clone01",
   "net0" : "virtio=BC:24:11:82:DA:53,bridge=vmbr0,firewall=1",
   "numa" : 0,
   "ostype" : "l26",
   "scsi0" : "local-btrfs:101/base-101-disk-0.raw,iothread=1,size=32G",
   "scsihw" : "virtio-scsi-single",
   "smbios1" : "uuid=2726d7d5-14b1-4ac4-a205-8ae3e932c083",
   "sockets" : 1,
   "template" : 1,
   "vga" : "virtio",
   "vmgenid" : "9447c8ee-d30a-468b-ae59-96d387ab5eda"
}
```


`pvesh  get /cluster/tasks --output-format=json-pretty | jq '[.[]| select(.type == "qmtemplate" )]'`

``` json
[
  {
    "endtime": 1719732092,
    "id": "101",
    "node": "anode",
    "saved": "1",
    "starttime": 1719732092,
    "status": "OK",
    "type": "qmtemplate",
    "upid": "UPID:anode:00004F98:000EA772:6681077C:qmtemplate:101:root@pam:",
    "user": "root@pam"
  }
]
```

 
 

### **tags sorting fun:**

setup:

each "packer" vm-id has 2 tags
- `packer`
- `packer-$( date --iso=s )`  ## so this turns into something like `packer-2024-06-30T01:39:28-07:00`


then the TF magic (sorry i did not really beautify this) 

``` hcl
data "proxmox_virtual_environment_vms" "packer_vms" {
  tags      = ["packer"]
}


locals {
  ## creates a new map with names of format "packer-$(date --iso=s)-<vm-id>", the "merge" part takes the first tag matching "packer-" and adds it to the payload as "packer_date" for visibility. then the "latest_packer_vm" handles the sort and picks the latest ( got that lookup from https://stackoverflow.com/questions/74153592/in-terraform-how-do-you-sort-a-list-of-objects-and-grab-the-last-value )
  ## note if 2x dates match, it does append the vm-id to the end, and that sort uses that vm-id too, so we assume in this case the vm-id is always greater for newer items

  ## areas to improve: regex matching packer-time could be more explicit; use try around areas; throw away the vm-id out of the "name" in "packer_vms" (maybe use `date --iso=ns`)

  packer_vms = { for a in data.proxmox_virtual_environment_vms.packer_vms.vms : "${[for b in a.tags : b if length(regexall("^packer-", b)) > 0][0]}-${a.vm_id}" => merge(a, { packer_date = [for b in a.tags : b if length(regexall("^packer-", b)) > 0][0] }) }

  latest_packer_vm = lookup(local.packer_vms, element(sort(keys(local.packer_vms)), length(local.packer_vms) - 1))
}


output "packer_vms_latest" {
  value = local.latest_packer_vm
}
```

**Comment by vanillaSprinkles** [comment]
you're welcome! dont consider it too minor - a lot of the aws-provider items are years worth of community driven development and have spoiled us in their ins/outs

but to be fair i went down the api-search rabbithole before trying @bpg's tag idea (then wish i tried the tags bit first, lol); maybe the down-the-road enhancement can be a fuzzy/glob match for the that data-object

**Comment by konstantin-kornienko** [comment]
Can add my 5 cents with the template VM issue. We're getting error when our terraform code is being run simultaneously. Here's the scenario:

Timing 0:
There's a template VM with tags ['packer', 'debian']

Timing 1:
Terraform tries to find the template to clone. It finds it by tags ['packer', 'debian']
There's a template VM with tags ['packer', 'debian']

Timing 2:
Terraform clones the template. For some time we have:
  *  template VM with tags ['packer', 'debian']
  * new VM with the same tags ['packer', 'debian'], because tags are cloned too. Even if you provide other tags to the VM resource, for several seconds they are not yet applied.

Timing 3:
Another terraform instance tries to fine the template to clone. It finds two VM by tags:
  * Template VM
  * Newly created VM by previous terraform instance.

In our case we're receiving the error:
error waiting for VM clone: All attempts fail:
│ #1: error cloning VM: received an HTTP 500 response - Reason: Linked clone feature is not supported for drive 'virtio0'

because newly created VM is found as a clone source instead of the template.


To overcome the issue, I'd suggest:
* Add `template` to the data source `virtual_environment_vms` (to both arguments and attributes). It will help us to filter our non-template VMs, like:

```
data "proxmox_virtual_environment_vms" "template" {
  tags      = ["packer", "template"]
  node_name = "..."
  template = true
}
```

**Comment by konstantin-kornienko** [comment]
Created a feature request: https://github.com/bpg/terraform-provider-proxmox/issues/1422

**Comment by vanillaSprinkles** [comment]
@konstantin-kornienko  "terraform code is being run simultaneously" hmm that presents a problem to me, nice catck. so to summarize: when scaling with multiple workspaces and querying from a common template and cloning, there's basically a race condition where the cloned-template (now a new vm-id) has the templates-tags, and is consumed by a clone-attempt on a tf-apply ran slightly after a different one.

I wonder then, in my tf-fun above then we can use the VM-ID as a workaround, if it matched and has more than one element, get the lowest-vm-id and assume that is the template. the problem here is earlier vm-ids can be deleted and the template could have a higher numeric value than the cloned vm. (lots of hcl-hackery when the api could help us out a bit more). i wonder if a custom "wait" can be added if more than 1 data-object is detected, and to retry with some exponential backoff... (maybe like a local-exec to a shell script that handles that? - would be a bit hairy but maybe a quick-n-dirty way to unblock if need it urgently)

almost feels like we need to have the proxmox folks add more support for more attributes in some of the api requests too; would make programming around it a bit easier. (especially trying to find out when a vm-template was created, vs having to lookup and filter the cluster/tasks (provided i did not miss the obvious) )

**Comment by konstantin-kornienko** [comment]
@vanillaSprinkles , thank you for sharing your thoughts!
I'd personally prefer not to implement hacks and hard code Template VM IDs in our automation code.

Trying to resolve our race-condition issue by extending the data source resource `vms` with template-only filtering:
https://github.com/bpg/terraform-provider-proxmox/pull/1423

It would fully resolve our specific use case.
